File: kavach-node/npm/src/mcp.ts
/**
 * Kavach middleware for MCP TypeScript servers.
 *
 * Wraps the official @modelcontextprotocol/sdk tool handlers
 * so every call passes through the Rust gate.
 *
 * @example
 * ```typescript
 * import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
 * import { Gate, McpKavachMiddleware, InMemorySessionStore } from 'kavach';
 *
 * const gate = Gate.fromFile('kavach.toml');
 * const kavach = new McpKavachMiddleware(gate, {
 *   sessionStore: new InMemorySessionStore(),
 * });
 *
 * const server = new McpServer({ name: 'my-server', version: '1.0' });
 *
 * server.tool('issue_refund', { amount: z.number(), orderId: z.string() },
 *   async ({ amount, orderId }) => {
 *     // Gate the call, throws KavachRefused if blocked,
 *     // KavachInvalidated if the session has been revoked.
 *     kavach.checkToolCall('issue_refund', { amount, orderId }, {
 *       callerId: 'agent-bot',
 *       callerKind: 'agent',
 *       sessionId: 'sess-42',
 *       roles: ['support'],
 *     });
 *
 *     // If we reach here, the gate permitted
 *     const result = await processRefund(orderId, amount);
 *     return { content: [{ type: 'text', text: JSON.stringify(result) }] };
 *   }
 * );
 * ```
 */

import {
  Gate,
  KavachInvalidated,
  type PrincipalKind,
  type EvaluateOptions,
  type GeoLocationInput,
} from './index';
import type { VerdictResult } from 'kavach-engine';

export interface McpCallerInfo {
  callerId: string;
  callerKind?: PrincipalKind;
  roles?: string[];
  sessionId?: string;
  ip?: string;
  /**
   * Current geographic location. Pair with `originGeo` and a
   * tolerant-mode `GeoLocationDrift` evaluator to tolerate
   * same-country IP hops.
   */
  currentGeo?: GeoLocationInput;
  /** Geographic location captured at session start. */
  originGeo?: GeoLocationInput;
}

/**
 * Pluggable per-session invalidation store.
 *
 * The MCP middleware calls `isInvalidated(sessionId)` before every
 * gated tool call so a peer-revoked session is refused without
 * reaching the gate, and `invalidate(sessionId)` on
 * `invalidateSession(...)` so the revocation fans out.
 *
 * The default backend is `InMemorySessionStore` (process-local
 * `Set`). Multi-replica deployments should swap in a Redis-backed
 * implementation: the interface is narrow enough that any
 * async-capable store fits without changes to the middleware.
 */
export interface SessionStore {
  isInvalidated(sessionId: string): boolean | Promise<boolean>;
  invalidate(sessionId: string): void | Promise<void>;
}

/**
 * Process-local session store, default backend for
 * `McpKavachMiddleware`. Tracks revoked session ids in a `Set`.
 * All operations are synchronous; the interface signature allows
 * Promise returns so Redis-backed stores slot in without changes
 * to the middleware.
 */
export class InMemorySessionStore implements SessionStore {
  private revoked = new Set<string>();

  isInvalidated(sessionId: string): boolean {
    return this.revoked.has(sessionId);
  }

  invalidate(sessionId: string): void {
    this.revoked.add(sessionId);
  }

  /** Number of revoked session ids currently held (observability / tests). */
  get size(): number {
    return this.revoked.size;
  }
}

export interface McpMiddlewareOptions {
  /**
   * Pluggable session store used for cross-replica invalidation
   * fan-out. Defaults to a fresh {@link InMemorySessionStore}.
   */
  sessionStore?: SessionStore;
}

/**
 * Kavach middleware for MCP servers.
 * All evaluation runs in the compiled Rust engine.
 */
export class McpKavachMiddleware {
  private gate: Gate;
  private store: SessionStore;
  private sessions = new Map<string, { actionCount: number }>();

  constructor(gate: Gate, options: McpMiddlewareOptions = {}) {
    this.gate = gate;
    this.store = options.sessionStore ?? new InMemorySessionStore();
  }

  /**
   * Check a tool call against the gate. Throws if blocked.
   *
   * When the call carries a `sessionId`, the middleware fast-paths
   * a session-store `isInvalidated` check BEFORE building the gate
   * context, a peer-revoked session surfaces as a
   * `KavachInvalidated` with `evaluator = "session_store"`.
   *
   * @param toolName - The MCP tool being called
   * @param params - Tool parameters (numeric values checked against invariants)
   * @param caller - Caller identity information
   * @throws KavachRefused if the gate blocks the action
   * @throws KavachInvalidated if the session is revoked (either via
   *   session-store fast-path or by a drift evaluator inside the gate)
   */
  async checkToolCall(
    toolName: string,
    params: Record<string, unknown>,
    caller: McpCallerInfo,
  ): Promise<void> {
    await this.fastPathInvalidationCheck(caller.sessionId);
    this.gate.check(this.buildContext(toolName, params, caller));
    this.recordAction(caller.sessionId);
  }

  /**
   * Evaluate a tool call without throwing. Returns the Verdict.
   *
   * If the session has been invalidated through the session store,
   * this returns a synthesised Invalidate verdict with
   * `evaluator = "session_store"`, the gate's own evaluator chain
   * is not run in that case, matching the Python middleware's
   * fast-path behaviour.
   */
  async evaluateToolCall(
    toolName: string,
    params: Record<string, unknown>,
    caller: McpCallerInfo,
  ): Promise<VerdictResult> {
    if (caller.sessionId && (await this.store.isInvalidated(caller.sessionId))) {
      return {
        kind: 'invalidate',
        evaluator: 'session_store',
        reason: `session ${caller.sessionId} has been invalidated`,
        code: null,
        tokenId: null,
        isPermit: false,
        isRefuse: false,
        isInvalidate: true,
        permitToken: null,
        signature: null,
      } as unknown as VerdictResult;
    }
    const verdict = this.gate.evaluate(this.buildContext(toolName, params, caller));
    if (verdict.isPermit) {
      this.recordAction(caller.sessionId);
    }
    return verdict;
  }

  /**
   * Create a wrapper function that gates any tool handler.
   *
   * @example
   * ```typescript
   * const guardedRefund = kavach.guardTool(
   *   'issue_refund',
   *   async (params) => processRefund(params),
   *   { callerId: 'agent-bot', callerKind: 'agent' },
   * );
   *
   * // Later:
   * const result = await guardedRefund({ amount: 500, orderId: 'ORD-123' });
   * ```
   */
  guardTool<T>(
    toolName: string,
    handler: (params: Record<string, unknown>) => T | Promise<T>,
    defaultCaller: McpCallerInfo,
  ): (params: Record<string, unknown>, callerOverride?: Partial<McpCallerInfo>) => Promise<T> {
    return async (params, callerOverride) => {
      const caller = { ...defaultCaller, ...callerOverride };
      await this.checkToolCall(toolName, params, caller);
      return handler(params);
    };
  }

  /**
   * Revoke a session across every replica that shares the
   * configured session store. Subsequent `checkToolCall` /
   * `evaluateToolCall` on the same session id will be refused
   * at the session-store fast-path.
   */
  async invalidateSession(sessionId: string): Promise<void> {
    await this.store.invalidate(sessionId);
  }

  private async fastPathInvalidationCheck(sessionId?: string): Promise<void> {
    if (!sessionId) return;
    if (await this.store.isInvalidated(sessionId)) {
      throw new KavachInvalidated(
        `session ${sessionId} has been invalidated`,
        'session_store',
      );
    }
  }

  private buildContext(
    toolName: string,
    params: Record<string, unknown>,
    caller: McpCallerInfo,
  ): EvaluateOptions {
    const numericParams: Record<string, number> = {};
    for (const [k, v] of Object.entries(params)) {
      if (typeof v === 'number') {
        numericParams[k] = v;
      }
    }

    return {
      principalId: caller.callerId,
      principalKind: caller.callerKind ?? 'agent',
      actionName: toolName,
      roles: caller.roles,
      params: Object.keys(numericParams).length > 0 ? numericParams : undefined,
      ip: caller.ip,
      sessionId: caller.sessionId,
      currentGeo: caller.currentGeo,
      originGeo: caller.originGeo,
    };
  }

  private recordAction(sessionId?: string): void {
    if (!sessionId) return;
    const session = this.sessions.get(sessionId) ?? { actionCount: 0 };
    session.actionCount++;
    this.sessions.set(sessionId, session);
  }
}

File: kavach-py/python/kavach/mcp.py
"""Kavach middleware for Python MCP servers.

Wraps MCP tool handlers so every call passes through the Rust gate.
Works with the official MCP Python SDK.

Example:
    from mcp.server import Server
    from kavach import Gate, McpKavachMiddleware

    gate = Gate.from_file("kavach.toml")
    kavach = McpKavachMiddleware(gate)

    server = Server("my-server")

    @server.call_tool()
    async def handle_tool(name: str, arguments: dict) -> list:
        # Gate the call — raises Refused if blocked
        kavach.check_tool_call(
            tool_name=name,
            params=arguments,
            caller_id="agent-bot",
            caller_kind="agent",
        )

        # If we reach here, the gate permitted
        if name == "issue_refund":
            return [TextContent(text=process_refund(arguments))]

Distributed deployments
-----------------------
Pass a shared session store to make ``invalidate_session`` visible
across service replicas::

    from kavach import RedisSessionStore

    store = RedisSessionStore.from_url("redis://localhost/0")
    kavach = McpKavachMiddleware(gate, session_store=store)

The middleware will consult the store before every gated call and
raise :class:`~kavach.Invalidated` when a peer has revoked the
session. Without ``session_store``, an in-process
:class:`~kavach.InMemorySessionStore` is used — same code path,
single-node scope.
"""

from dataclasses import dataclass, field
from typing import Union

from kavach._kavach_engine import (
    ActionContext,
    GeoLocation,
    InMemorySessionStore,
    RedisSessionStore,
)
from kavach.wrappers import Gate, Invalidated


SessionStoreBackend = Union[InMemorySessionStore, RedisSessionStore]


@dataclass
class McpSession:
    """Tracks state for an MCP session."""

    session_id: str
    caller_id: str
    caller_kind: str = "agent"
    roles: list[str] = field(default_factory=list)
    ip: str | None = None
    action_count: int = 0
    invalidated: bool = False


class McpKavachMiddleware:
    """Kavach middleware for MCP servers.

    All evaluation runs in the compiled Rust engine.
    This class manages sessions and translates MCP concepts
    to Kavach ActionContext.
    """

    def __init__(
        self,
        gate: Gate,
        *,
        session_store: SessionStoreBackend | None = None,
    ):
        """Create a middleware bound to a :class:`~kavach.Gate`.

        Args:
            gate: The Kavach gate this middleware evaluates against.
            session_store: Optional backend for cross-replica session
                state — accepts an :class:`~kavach.InMemorySessionStore`
                or a :class:`~kavach.RedisSessionStore`. When unset, an
                in-process ``InMemorySessionStore`` is used internally
                so the same code path serves both single-node and
                multi-node deployments. When set to a Redis-backed
                store, :meth:`invalidate_session` fans out to peer
                replicas and :meth:`check_tool_call` honours
                cross-node invalidations before reaching the gate.
        """
        self._gate = gate
        self._store: SessionStoreBackend = (
            session_store if session_store is not None else InMemorySessionStore()
        )
        # Local cache of McpSession views (principal-facing fields that
        # aren't part of SessionState). Action-count mutations in here
        # are observational only — the gate itself reads action_count
        # from the ActionContext the caller passes in, not from this
        # cache.
        self._sessions: dict[str, McpSession] = {}

    def check_tool_call(
        self,
        *,
        tool_name: str,
        params: dict,
        caller_id: str,
        caller_kind: str = "agent",
        roles: list[str] | None = None,
        session_id: str | None = None,
        ip: str | None = None,
        current_geo: GeoLocation | None = None,
        origin_geo: GeoLocation | None = None,
    ) -> None:
        """Check a tool call against the gate. Raises if blocked.

        Args:
            tool_name: The MCP tool being called.
            params: Tool parameters.
            caller_id: Identity of the caller.
            caller_kind: "agent", "user", or "service".
            roles: Caller's roles.
            session_id: Session ID for drift tracking.
            ip: Caller's IP address.
            current_geo: Geographic location at this call. Set this plus
                ``origin_geo`` to unlock tolerant-mode GeoLocationDrift
                (an IP change within ``max_distance_km`` downgrades to a
                warning instead of a violation).
            origin_geo: Geographic location at session start.

        Raises:
            kavach.Refused: If the gate blocks the action.
            kavach.Invalidated: If the session is revoked (locally or
                by a peer via the shared session store).
        """
        # Fast-path remote-invalidation check. A peer replica that called
        # `invalidate_session` will have flipped the shared store's flag;
        # we honour that before even building the ActionContext.
        if session_id and self._store.is_invalidated(session_id):
            raise Invalidated(
                f"session {session_id} invalidated by a peer node",
                "session_store",
            )

        # Build numeric params for invariant checks
        numeric_params = {
            k: float(v) for k, v in params.items() if isinstance(v, (int, float))
        }

        ctx = ActionContext(
            principal_id=caller_id,
            principal_kind=caller_kind,
            action_name=tool_name,
            roles=roles or [],
            params=numeric_params or None,
            ip=ip,
            session_id=session_id,
            current_geo=current_geo,
            origin_geo=origin_geo,
        )

        # Crosses into Rust for evaluation
        self._gate.check(ctx)

        # If we get here, gate permitted — track the action locally
        if session_id:
            session = self._sessions.setdefault(
                session_id,
                McpSession(
                    session_id=session_id,
                    caller_id=caller_id,
                    caller_kind=caller_kind,
                    roles=roles or [],
                    ip=ip,
                ),
            )
            session.action_count += 1

    def evaluate_tool_call(
        self,
        *,
        tool_name: str,
        params: dict,
        caller_id: str,
        caller_kind: str = "agent",
        roles: list[str] | None = None,
        session_id: str | None = None,
        ip: str | None = None,
        current_geo: GeoLocation | None = None,
        origin_geo: GeoLocation | None = None,
    ):
        """Evaluate without raising — returns the Verdict object.

        Use this when you want to handle the verdict yourself
        rather than using exception flow. See ``check_tool_call`` for
        the semantics of ``current_geo`` / ``origin_geo``.
        """
        numeric_params = {
            k: float(v) for k, v in params.items() if isinstance(v, (int, float))
        }

        ctx = ActionContext(
            principal_id=caller_id,
            principal_kind=caller_kind,
            action_name=tool_name,
            roles=roles or [],
            params=numeric_params or None,
            ip=ip,
            session_id=session_id,
            current_geo=current_geo,
            origin_geo=origin_geo,
        )

        return self._gate.evaluate(ctx)

    def get_session(self, session_id: str) -> McpSession | None:
        """Return the locally-cached :class:`McpSession`, updated for
        remote invalidation from the shared store.
        """
        session = self._sessions.get(session_id)
        if session is None:
            return None
        # Reflect remote invalidation in the local view so callers that
        # only read this object still see the correct state.
        if not session.invalidated and self._store.is_invalidated(session_id):
            session.invalidated = True
        return session

    def invalidate_session(self, session_id: str) -> None:
        """Invalidate locally and fan out via the shared session store.

        Local caches are updated immediately. The shared-store write
        is what makes the invalidation visible to peer replicas —
        their next :meth:`check_tool_call` will see it and raise
        :class:`~kavach.Invalidated`.
        """
        if session_id in self._sessions:
            self._sessions[session_id].invalidated = True
        self._store.invalidate(session_id)
